Support parallel environment builds

tldahlgren opened the pull request:
Replaces #15415
Closes #16724.

As of #13100, Spack installs the dependencies of a _single_ spec in parallel.  This PR extends that support to multiple specifications as found in environment builds.  

The specs and kwargs for each *uninstalled* package (when not force-replacing installations) of an environment are collected, passed to the PackageInstaller, and processed using a single build queue.

Note: A locking issue was detected prior to commit (`58b1036`) starting an environment build from two separate processes at the same time.  Restoration of the process of skipping already installed packages seemed to alleviate this problem.

TODO

- [x] Finish updating the unit tests based on `PackageInstaller`'s use of `BuildRequest` and the associated changes
- [x] Change `environment.py`'s `install_all` to use the `PackageInstaller` directly
- [x] Change the `install` command to leverage the new installation process for multiple specs
- [x] Resolve test failures
- [x] Change install output messages for external packages (e.g., `[+] /usr` to `[+] /usr (external bzip2-1.0.8-<dag-hash>`)
- [x] Fix incomplete environment install's view setup/update and not confirming all packages are installed (?)
- [x] Ensure externally installed package dependencies are properly accounted for in remaining build tasks
~- [ ] Address coordination issues between multiple environment `spack install`s.~
- [x] Add tests for coverage (if insufficient and can identity the appropriate, uncovered non-comment lines)
- [x] Add documentation
- [x] Resolve multi-compiler environment install issues
- [x] Fix issue with environment installation reporting (thanks `matz-e`)

tldahlgren committed:
Changed installer approach to identifying explicit packages

tldahlgren committed:
Fix docstring wording

tldahlgren committed:
Remove spec property from installer classes.

tldahlgren committed:
Refactor to reduce Installer's reliance on package/spec attributes

tldahlgren committed:
flake8 fix and comment tweak.

tldahlgren committed:
Added BuildRequest and eliminated priority queue dependence on package attrs

tldahlgren commented:
While the unit tests still need to be updated, I was able to successfully build `m4` with `spack install m4` at commit a904f3d.

tldahlgren committed:
Updated install tests

tldahlgren committed:
Added BuildRequest unit tests

tldahlgren committed:
Resolved dev_build test failure

tldahlgren committed:
Fixed flake8 errors

tldahlgren committed:
Cleaned up install exception handling for single vs. multiple specs

tldahlgren committed:
Refactored create_installer and added multi-spec install tests

tldahlgren committed:
Initial pass over environment's install_all use of PackageInstaller

tldahlgren committed:
Finished preliminary pass to parallelize multi-spec builds

tldahlgren committed:
Resolved unit test issues

tldahlgren committed:
Fixed flake8 issues in test

tldahlgren committed:
Fix install message for external packages; ensure install confirms all specs installed

tldahlgren committed:
Flag externals and upstream as installed

tldahlgren committed:
Ensure external and upstream packages properly tracked as installed

tldahlgren committed:
Added environment failure reproducer and improved solution

tldahlgren commented:
@ping becker33

tldahlgren committed:
Updated docs associated with Installing messages

tldahlgren committed:
Added install-level build parallelism subsection

tldahlgren committed:
Fix docs link issue

matz-e commented:
Awesome!

Short question: will this attempt to re-install failed dependencies if they occur multiple times in the environments forest of DAGs, or cache failures and skip them/dependencies?

tldahlgren commented:
> Awesome!

Thanks.
 

> Short question: will this attempt to re-install failed dependencies if they occur multiple times in the environments forest of DAGs, or cache failures and skip them/dependencies?

No it will not attempt to re-install failed dependencies while processing the concurrent installs.  All dependents of the failed package will "fail" (i.e., be skipped), with a ripple effect on ancestors/dependents.

The idea is to give you a chance to resolve the failure issue(s) before retrying the installation.

tldahlgren committed:
Traverse dependencies based on request's options

matz-e commented:
> No it will not attempt to re-install failed dependencies while processing the concurrent installs. All dependents of the failed package will "fail" (i.e., be skipped), with a ripple effect on ancestors/dependents.

Great!

I just got a chance to test this, and managed to break things with the following error:
```
==> [2020-09-30-14:12:28.417687] Error: Cannot proceed with isl-0.20-ij2vhzkelvc6i2e47symeotc3zdsgics: 1 uninstalled dependency: gmp-6.1.2-raqmdaosldybjrn6y2xnnnyy2f63hg6b
Traceback (most recent call last):
  File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/main.py", line 762, in main
    return _invoke_command(command, parser, args, unknown)
  File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/main.py", line 490, in _invoke_command
    return_val = command(parser, args)
  File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/cmd/install.py", line 297, in install
    env.install_all(args, **kwargs)
  File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/environment.py", line 1291, in install_all
    builder.install()
  File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/installer.py", line 1480, in install
    ','.join(task.uninstalled_deps)))
spack.installer.InstallError: Cannot proceed with isl-0.20-ij2vhzkelvc6i2e47symeotc3zdsgics: 1 uninstalled dependency: gmp-6.1.2-raqmdaosldybjrn6y2xnnnyy2f63hg6b
```
With the following environment (condensed):
```
spack:
  view: false
  concretization: separately
  packages:
    all:
      compiler: [gcc@4.8.5]
      target: []
      version: []
      buildable: true
      providers: {}
  specs:
  - gcc@9.3.0
  - gcc@10.1.0
```

More details / logs in https://gist.github.com/matz-e/0c494010f9de9b8e42b63353169c7a21#file-spack-yaml

The installation will fail, proceed a bit more when restarted, and then keep giving me the same error. I poked around a little bit, was wondering if this is due to the separate concretization?

There are a lot of references to `spec.dependents`, where `spec` is fetched by reference to the `pkg_id`. Seems like if the same concrete dependent spec is observed in different requested root specs, there may only be a partial set of dependents referenced like that? (Two identical `pkg_id`s with different set of `spec.denpendents`)

tldahlgren commented:
> > No it will not attempt to re-install failed dependencies while processing the concurrent installs. All dependents of the failed package will "fail" (i.e., be skipped), with a ripple effect on ancestors/dependents.
> 
> Great!
> 
> I just got a chance to test this, and managed to break things with the following error:
> 
> ```
> ==> [2020-09-30-14:12:28.417687] Error: Cannot proceed with isl-0.20-ij2vhzkelvc6i2e47symeotc3zdsgics: 1 uninstalled dependency: gmp-6.1.2-raqmdaosldybjrn6y2xnnnyy2f63hg6b
> Traceback (most recent call last):
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/main.py", line 762, in main
>     return _invoke_command(command, parser, args, unknown)
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/main.py", line 490, in _invoke_command
>     return_val = command(parser, args)
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/cmd/install.py", line 297, in install
>     env.install_all(args, **kwargs)
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/environment.py", line 1291, in install_all
>     builder.install()
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/installer.py", line 1480, in install
>     ','.join(task.uninstalled_deps)))
> spack.installer.InstallError: Cannot proceed with isl-0.20-ij2vhzkelvc6i2e47symeotc3zdsgics: 1 uninstalled dependency: gmp-6.1.2-raqmdaosldybjrn6y2xnnnyy2f63hg6b
> ```
> 
> With the following environment (condensed):
> 
> ```
> spack:
>   view: false
>   concretization: separately
>   packages:
>     all:
>       compiler: [gcc@4.8.5]
>       target: []
>       version: []
>       buildable: true
>       providers: {}
>   specs:
>   - gcc@9.3.0
>   - gcc@10.1.0
> ```
> 
> More details / logs in https://gist.github.com/matz-e/0c494010f9de9b8e42b63353169c7a21#file-spack-yaml
> 
> The installation will fail, proceed a bit more when restarted, and then keep giving me the same error. I poked around a little bit, was wondering if this is due to the separate concretization?
> 
> There are a lot of references to `spec.dependents`, where `spec` is fetched by reference to the `pkg_id`. Seems like if the same concrete dependent spec is observed in different requested root specs, there may only be a partial set of dependents referenced like that? (Two identical `pkg_id`s with different set of `spec.denpendents`)

Thanks for trying it and letting me know the issue.

tldahlgren commented:
> > No it will not attempt to re-install failed dependencies while processing the concurrent installs. All dependents of the failed package will "fail" (i.e., be skipped), with a ripple effect on ancestors/dependents.
> 
> Great!
> 
> I just got a chance to test this, and managed to break things with the following error:
> 
> ```
> ==> [2020-09-30-14:12:28.417687] Error: Cannot proceed with isl-0.20-ij2vhzkelvc6i2e47symeotc3zdsgics: 1 uninstalled dependency: gmp-6.1.2-raqmdaosldybjrn6y2xnnnyy2f63hg6b
> Traceback (most recent call last):
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/main.py", line 762, in main
>     return _invoke_command(command, parser, args, unknown)
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/main.py", line 490, in _invoke_command
>     return_val = command(parser, args)
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/cmd/install.py", line 297, in install
>     env.install_all(args, **kwargs)
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/environment.py", line 1291, in install_all
>     builder.install()
>   File "/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/installer.py", line 1480, in install
>     ','.join(task.uninstalled_deps)))
> spack.installer.InstallError: Cannot proceed with isl-0.20-ij2vhzkelvc6i2e47symeotc3zdsgics: 1 uninstalled dependency: gmp-6.1.2-raqmdaosldybjrn6y2xnnnyy2f63hg6b
> ```
> 
> With the following environment (condensed):
> 
> ```
> spack:
>   view: false
>   concretization: separately
>   packages:
>     all:
>       compiler: [gcc@4.8.5]
>       target: []
>       version: []
>       buildable: true
>       providers: {}
>   specs:
>   - gcc@9.3.0
>   - gcc@10.1.0
> ```
> 
> More details / logs in https://gist.github.com/matz-e/0c494010f9de9b8e42b63353169c7a21#file-spack-yaml
> 
> The installation will fail, proceed a bit more when restarted, and then keep giving me the same error. I poked around a little bit, was wondering if this is due to the separate concretization?
> 
> There are a lot of references to `spec.dependents`, where `spec` is fetched by reference to the `pkg_id`. Seems like if the same concrete dependent spec is observed in different requested root specs, there may only be a partial set of dependents referenced like that? (Two identical `pkg_id`s with different set of `spec.denpendents`)

I experienced similar issues concretizing separately and think I've addressed them. 

@matz-e I ran your `spack.yaml` using the current commit (`80a8816`) and it was able to install the `isl` package.  (It is still running for me btw.)  Would you mind trying again with this commit and letting me know how it goes?

UPDATE: The latest version installed `isl-0.21-hsrneph7vglanxas3cif5eiukpp53cdm` for me BUT not `isl-0.20-ij2vhzkelvc6i2e47symeotc3zdsgics` so I will need to look at this more closely to come up with more than a trivial fix.

tldahlgren committed:
Distinguish between initializing and replacing an installer BuildTask

tldahlgren committed:
Fix tracking of all dependents across multiple install lspecs

matz-e commented:
Thanks, these latest commits fix the dependency issue. I've kicked off my little test case, and with 3 build processes, I see only one building things, though :(

I checked the debug output of one of them, and see this:
```
==> [2020-10-14-14:14:36.853204] Acquiring a write lock on zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d with timeout 1e-09
==> [2020-10-14-14:14:36.853975] Failed to acquire a write lock for zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d due to LockTimeoutError: Timed out waiting for a write lock.
==> [2020-10-14-14:14:36.854064] Acquiring a read lock on zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d with timeout 3
==> [2020-10-14-14:14:36.855449] Creating stage lock spack-stage-zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d
==> [2020-10-14-14:14:36.866201] Flagging zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d as installed
==> [2020-10-14-14:14:36.866312] Removing zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from gcc-10.1.0-dgfk5jdu63hadohn4injud4ue2ct6r7y's uninstalled dependencies.
==> [2020-10-14-14:14:36.866402] gcc-10.1.0-dgfk5jdu63hadohn4injud4ue2ct6r7y: Removed zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from uninstalled deps list: {'mpfr-4.0.2-yfoprebczoqpokivqgmnjn5os4r7mxw6', 'isl-0.21-hsrneph7vglanxas3cif5eiukpp53cdm', 'mpc-1.1.0-ijex5qs7pt44n7whqlgwqj4jwc3czkac', 'zstd-1.4.5-zw7esm7zwzxsrlsperpuzx6opq4gp4hu'}
==> [2020-10-14-14:14:36.866471] Removing zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo's uninstalled dependencies.
==> [2020-10-14-14:14:36.866532] gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo: Removed zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from uninstalled deps list: set()
==> [2020-10-14-14:14:36.866585] Removing zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from zstd-1.4.5-zw7esm7zwzxsrlsperpuzx6opq4gp4hu's uninstalled dependencies.
==> [2020-10-14-14:14:36.866642] zstd-1.4.5-zw7esm7zwzxsrlsperpuzx6opq4gp4hu: Removed zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from uninstalled deps list: set()
[+] /gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/opt/spack/linux-rhel7-haswell/gcc-4.8.5/zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d
==> [2020-10-14-14:14:36.867940] Acquiring a write lock on gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo with timeout 1e-09
==> [2020-10-14-14:14:36.868829] Failed to acquire a write lock for gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo due to LockTimeoutError: Timed out waiting for a write lock.
==> [2020-10-14-14:14:36.868931] Acquiring a read lock on gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo with timeout 3
==> [2020-10-14-14:14:39.874490] Failed to acquire a read lock for gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo due to LockTimeoutError: Timed out waiting for a read lock.
==> [2020-10-14-14:14:39.874711] Installing gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo in progress by another process
```
where I'm in a state where one process builds `gcc@9.3.0`, and two others are attempting to also build that `gcc@9.3.0`. But `zstd@1.4.5` seems to have an empty set of dependencies and could already be built, too? Am I reading this right? None of my two "available" build processes attempted to build it…

tldahlgren commented:
> Thanks, these latest commits fix the dependency issue. I've kicked off my little test case, and with 3 build processes, I see only one building things, though :(
> 
> I checked the debug output of one of them, and see this:
> 
> ```
> ==> [2020-10-14-14:14:36.853204] Acquiring a write lock on zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d with timeout 1e-09
> ==> [2020-10-14-14:14:36.853975] Failed to acquire a write lock for zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d due to LockTimeoutError: Timed out waiting for a write lock.
> ==> [2020-10-14-14:14:36.854064] Acquiring a read lock on zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d with timeout 3
> ==> [2020-10-14-14:14:36.855449] Creating stage lock spack-stage-zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d
> ==> [2020-10-14-14:14:36.866201] Flagging zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d as installed
> ==> [2020-10-14-14:14:36.866312] Removing zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from gcc-10.1.0-dgfk5jdu63hadohn4injud4ue2ct6r7y's uninstalled dependencies.
> ==> [2020-10-14-14:14:36.866402] gcc-10.1.0-dgfk5jdu63hadohn4injud4ue2ct6r7y: Removed zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from uninstalled deps list: {'mpfr-4.0.2-yfoprebczoqpokivqgmnjn5os4r7mxw6', 'isl-0.21-hsrneph7vglanxas3cif5eiukpp53cdm', 'mpc-1.1.0-ijex5qs7pt44n7whqlgwqj4jwc3czkac', 'zstd-1.4.5-zw7esm7zwzxsrlsperpuzx6opq4gp4hu'}
> ==> [2020-10-14-14:14:36.866471] Removing zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo's uninstalled dependencies.
> ==> [2020-10-14-14:14:36.866532] gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo: Removed zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from uninstalled deps list: set()
> ==> [2020-10-14-14:14:36.866585] Removing zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from zstd-1.4.5-zw7esm7zwzxsrlsperpuzx6opq4gp4hu's uninstalled dependencies.
> ==> [2020-10-14-14:14:36.866642] zstd-1.4.5-zw7esm7zwzxsrlsperpuzx6opq4gp4hu: Removed zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d from uninstalled deps list: set()
> [+] /gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/opt/spack/linux-rhel7-haswell/gcc-4.8.5/zlib-1.2.11-zolwez4onex3iueybyju2zi55njtdg5d
> ==> [2020-10-14-14:14:36.867940] Acquiring a write lock on gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo with timeout 1e-09
> ==> [2020-10-14-14:14:36.868829] Failed to acquire a write lock for gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo due to LockTimeoutError: Timed out waiting for a write lock.
> ==> [2020-10-14-14:14:36.868931] Acquiring a read lock on gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo with timeout 3
> ==> [2020-10-14-14:14:39.874490] Failed to acquire a read lock for gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo due to LockTimeoutError: Timed out waiting for a read lock.
> ==> [2020-10-14-14:14:39.874711] Installing gcc-9.3.0-ukhvotjr7b27pfdsjekh34kc3qtg6bpo in progress by another process
> ```
> 
> where I'm in a state where one process builds `gcc@9.3.0`, and two others are attempting to also build that `gcc@9.3.0`. But `zstd@1.4.5` seems to have an empty set of dependencies and could already be built, too? Am I reading this right? None of my two "available" build processes attempted to build it…

Thank you for confirming the fix worked!

I too noticed issues with a multi-process build of the simple, two-compiler installation environment as noted in an update I made to this PRs description.  I will look into this more closely.

It is important to keep in mind that parallelism is effectively limited by the maximum number of packages that can be built at the same time, which itself is limited by the maximum number of packages with **no** uninstalled dependencies.  

IIRC  the two versions of `gcc` being installed shared common dependencies in their DAGS on the same versions of at least a couple of packages.  You can see the overlap if you compare the output of `spack spec gcc@9.3.0` to that of `spack spec gcc@10.1.0`.

tldahlgren committed:
Restore pre-screening specs in an environment for already being installed

tldahlgren commented:
Closing and re-opening to see if it resolves the fortran build issue.

becker33 commented on lib/spack/spack/cmd/install.py:
```diff
@@ -253,6 +260,11 @@ def install_spec(cli_args, kwargs, abstract_spec, spec):
+    finally:
+        if env:
+            with env.write_transaction():
+                env.regenerate_views()
```
Shouldn't `env.install_all` handle this?

tldahlgren replied:
Good question.  I don't know why it was originally within the `try-except` block so I retained it here.  But I will move it.

becker33 commented on lib/spack/spack/cmd/install.py:
```diff
@@ -385,13 +397,17 @@ def install(parser, args, **kwargs):
-                else:
-                    install_spec(args, kwargs, abstract, concrete)
-
+            if any([concrete in installed for concrete in specs]):
+                # Preserve ordering if force replacement of any installed specs
+                tty.debug('Performing serial installs due to forced '
+                          'replacement of installed concrete specs.')
+                for abstract, concrete in zip(abstract_specs, specs):
+                    if concrete in installed:
+                        with fs.replace_directory_transaction(concrete.prefix):
+                            install_specs(args, kwargs, [(abstract, concrete)])
+                    else:
+                        install_specs(args, kwargs, [(abstract, concrete)])
```
I think this can all be replaced by the logic in #15256 . That will avoid having any serial installs

tldahlgren replied:
Thanks for the PR number.  I held off changing this section after we talked about this before so I can integrate that code.

tldahlgren replied:
Merged with develop to pick this up.

tldahlgren replied:
> > I too noticed issues with a multi-process build of the simple, two-compiler installation environment as noted in an update I made to this PRs description. I will look into this more closely.
> > It is important to keep in mind that parallelism is effectively limited by the maximum number of packages that can be built at the same time, which itself is limited by the maximum number of packages with **no** uninstalled dependencies.
> > IIRC the two versions of `gcc` being installed shared common dependencies in their DAGS on the same versions of at least a couple of packages. You can see the overlap if you compare the output of `spack spec gcc@9.3.0` to that of `spack spec gcc@10.1.0`.
> 
> I'm aware that there's a lot of overlap, but I somehow could not get more than one process building any package at all. I just tested it by simply adding the 6-8 most recent versions of CMake to an environment (once dependencies are installed, these should all be orthogonal), and using just
> 
> ```
> spack -e debug_environment install -j 6
> ```
> 
> in 4 terminals. This has one process installing one CMake after another, while my other 3 keep saying that it is `in progress by another process`. _Maybe I'm using the command wrong?_
> 
> I added a breakpoint after re-queuing tasks if a lock cannot be obtained (around line 1540 in `installer.py`), and it seems to me that the same task is popped from the heap, which looks like:
> 
> ```
> > /gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/installer.py(1542)install()
> -> pdb.set_trace()
> (Pdb) pprint.pprint([(t[0], t[1].pkg_id) for t in self.build_pq if t[1].status != STATUS_REMOVED])
> [((0, 13), 'cmake-3.16.5-y2q7fnh5piydazslo44gwze526kmavka'),
>  ((0, 14), 'cmake-3.16.2-lcrrvvf2ghljvwu3flrssjzdzsvvaxn3'),
>  ((0, 15), 'cmake-3.16.1-ccldrcnz7mjguukpejljyq2lpq6vvxsi')]
> ```
> 
> and where `cmake@3.16.5` is re-queued after the lock fails, with the same priority and build sequence counter. Shouldn't the key of the heap account for lock fails to bump the priority (or other key component) up a bit and allow to rotate through all tasks with the same priority? It may well be that I'm just a little too dense for this, though.

Thanks for posting the issue.  I'll look at this more once I merge to develop.

matz-e commented:
> I too noticed issues with a multi-process build of the simple, two-compiler installation environment as noted in an update I made to this PRs description. I will look into this more closely.
> 
> It is important to keep in mind that parallelism is effectively limited by the maximum number of packages that can be built at the same time, which itself is limited by the maximum number of packages with **no** uninstalled dependencies.
> 
> IIRC the two versions of `gcc` being installed shared common dependencies in their DAGS on the same versions of at least a couple of packages. You can see the overlap if you compare the output of `spack spec gcc@9.3.0` to that of `spack spec gcc@10.1.0`.

I'm aware that there's a lot of overlap, but I somehow could not get more than one process building any package at all. I just tested it by simply adding the 6-8 most recent versions of CMake to an environment (once dependencies are installed, these should all be orthogonal), and using just
```
spack -e debug_environment install -j 6
```
in 4 terminals. This has one process installing one CMake after another, while my other 3 keep saying that it is `in progress by another process`. _Maybe I'm using the command wrong?_

I added a breakpoint after re-queuing tasks if a lock cannot be obtained (around line 1540 in `installer.py`), and it seems to me that the same task is popped from the heap, which looks like:
```
> /gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/installer.py(1542)install()
-> pdb.set_trace()
(Pdb) pprint.pprint([(t[0], t[1].pkg_id) for t in self.build_pq if t[1].status != STATUS_REMOVED])
[((0, 13), 'cmake-3.16.5-y2q7fnh5piydazslo44gwze526kmavka'),
 ((0, 14), 'cmake-3.16.2-lcrrvvf2ghljvwu3flrssjzdzsvvaxn3'),
 ((0, 15), 'cmake-3.16.1-ccldrcnz7mjguukpejljyq2lpq6vvxsi')]
```
and where `cmake@3.16.5` is re-queued after the lock fails, with the same priority and build sequence counter. Shouldn't the key of the heap account for lock fails to bump the priority (or other key component) up a bit and allow to rotate through all tasks with the same priority? It may well be that I'm just a little too dense for this, though.

tldahlgren committed:
Merge remote-tracking branch 'origin/develop' into features/parallel-env-builds

tldahlgren committed:
Fix two lines not updated in previous commit

tldahlgren committed:
Moved post-install regeneration views from install to environment

tldahlgren committed:
Fix failing test_env_install_same_spec_twice issue

tldahlgren commented:
> > I too noticed issues with a multi-process build of the simple, two-compiler installation environment as noted in an update I made to this PRs description. I will look into this more closely.
> > It is important to keep in mind that parallelism is effectively limited by the maximum number of packages that can be built at the same time, which itself is limited by the maximum number of packages with **no** uninstalled dependencies.
> > IIRC the two versions of `gcc` being installed shared common dependencies in their DAGS on the same versions of at least a couple of packages. You can see the overlap if you compare the output of `spack spec gcc@9.3.0` to that of `spack spec gcc@10.1.0`.
> 
> I'm aware that there's a lot of overlap, but I somehow could not get more than one process building any package at all. I just tested it by simply adding the 6-8 most recent versions of CMake to an environment (once dependencies are installed, these should all be orthogonal), and using just
> 
> ```
> spack -e debug_environment install -j 6
> ```
> 
> in 4 terminals. This has one process installing one CMake after another, while my other 3 keep saying that it is `in progress by another process`. _Maybe I'm using the command wrong?_
> 
> I added a breakpoint after re-queuing tasks if a lock cannot be obtained (around line 1540 in `installer.py`), and it seems to me that the same task is popped from the heap, which looks like:
> 
> ```
> > /gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/installer.py(1542)install()
> -> pdb.set_trace()
> (Pdb) pprint.pprint([(t[0], t[1].pkg_id) for t in self.build_pq if t[1].status != STATUS_REMOVED])
> [((0, 13), 'cmake-3.16.5-y2q7fnh5piydazslo44gwze526kmavka'),
>  ((0, 14), 'cmake-3.16.2-lcrrvvf2ghljvwu3flrssjzdzsvvaxn3'),
>  ((0, 15), 'cmake-3.16.1-ccldrcnz7mjguukpejljyq2lpq6vvxsi')]
> ```
> 
> and where `cmake@3.16.5` is re-queued after the lock fails, with the same priority and build sequence counter. Shouldn't the key of the heap account for lock fails to bump the priority (or other key component) up a bit and allow to rotate through all tasks with the same priority? It may well be that I'm just a little too dense for this, though.

This is great additional details.  Thanks!

Another locking issue has popped up elsewhere -- related to how we determine another process is installing the software, permissions, and lock errors -- that I'm looking at next.  It _may_ be related to this issue but it's hard to tell at this point.

@matz-e Which OS are you using?

tldahlgren commented:
> > and where `cmake@3.16.5` is re-queued after the lock fails, with the same priority and build sequence counter. Shouldn't the key of the heap account for lock fails to bump the priority (or other key component) up a bit and allow to rotate through all tasks with the same priority? It may well be that I'm just a little too dense for this, though.

I missed the counter when I refactored the `BuildTask`-related code.  Thank you for pointing that out!

tldahlgren committed:
Ensure requeued build tasks get new sequence number

matz-e commented:
> @matz-e Which OS are you using?

RedHat 7.6

> I missed the counter when I refactored the BuildTask-related code. Thank you for pointing that out!

Thanks for the fix! And for all the work on this! Got my test-case(s) running smoothly so far, and I'm eager to try this out in a large-scale deployment.

tldahlgren commented:
@becker33 I addressed your feedback and appear to have resolved matz-e's issues.

tldahlgren commented:
> > @matz-e Which OS are you using?
> 
> RedHat 7.6

Good to know.  Thanks.
> 
> > I missed the counter when I refactored the BuildTask-related code. Thank you for pointing that out!
> 
> Thanks for the fix! And for all the work on this! Got my test-case(s) running smoothly so far, and I'm eager to try this out in a large-scale deployment.

This is great news.  Thank you for testing this PR and keeping me posted on how it works for you.

matz-e commented:
> This is great news. Thank you for testing this PR and keeping me posted on how it works for you.

Thanks! I've managed to fix most of our build issues and get another trial-deployment going. With this PR folded in, we get a nice speed-up, and a full deployment takes probably around 1/3 of the time (it's a little hard to measure with the required restarts). I'm currently testing this with 12 concurrent processes (each in turn running with `-j 12`) on SLURM, occupying 2 of our compute nodes. Spack, the working directory, and the install directory are all on GPFS 5. I added a delay to my SLURM wrapper so that the build processes start staggered in time (otherwise I got failures that also look like race conditions).

Sporadically, I notice that a spec is attempted to be built twice and fails the whole installation (I suspect this may have to do something with locking and GPFS, since I'm using two nodes). Otherwise, I also noticed that the following environment:
```
spack:
  view: false
  concretization: separately
  packages:
    all:
      compiler: [gcc@4.8.5]
      providers: {}
      version: []
      buildable: true
      target: [x86_64]
  config:
    install_missing_compilers: true
  specs:
  - gcc@9.3.0
  - llvm@10.0.0%gcc@9.3.0
```
Fails with errors like:
```
==> Error: NoCompilerForSpecError: No compilers for operating system rhel7 satisfy spec gcc@9.3.0

/gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/package.py:1197, in compiler:
       1194        """Get the spack.compiler.Compiler object used to build this package"""
       1195        if not self.spec.concrete:
       1196            raise ValueError("Can only get a compiler for a concrete package.")
  >>   1197        return spack.compilers.compiler_for_spec(self.spec.compiler,
       1198                                                 self.spec.architecture)
```
I would probably solve this by factoring out LLVM and running it after a bootstrap stage to compile GCC.

tldahlgren commented:
> > This is great news. Thank you for testing this PR and keeping me posted on how it works for you.
> 
> Thanks! I've managed to fix most of our build issues and get another trial-deployment going. With this PR folded in, we get a nice speed-up, and a full deployment takes probably around 1/3 of the time (it's a little hard to measure with the required restarts). 

Glad to hear of the speed up!


>I'm currently testing this with 12 concurrent processes (each in turn running with `-j 12`) on SLURM, occupying 2 of our compute nodes. Spack, the working directory, and the install directory are all on GPFS 5. I added a delay to my SLURM wrapper so that the build processes start staggered in time (otherwise I got failures that also look like race conditions).

Good to know.  We've seen things like this before with building the provider cache the first time.

Do you have debug output from where the builds were hanging before you staggered the start time?

> 
> Sporadically, I notice that a spec is attempted to be built twice and fails the whole installation (I suspect this may have to do something with locking and GPFS, since I'm using two nodes). Otherwise, I also noticed that the following environment:
> 
> ```
> spack:
>   view: false
>   concretization: separately
>   packages:
>     all:
>       compiler: [gcc@4.8.5]
>       providers: {}
>       version: []
>       buildable: true
>       target: [x86_64]
>   config:
>     install_missing_compilers: true
>   specs:
>   - gcc@9.3.0
>   - llvm@10.0.0%gcc@9.3.0
> ```
> 
> Fails with errors like:
> 
> ```
> ==> Error: NoCompilerForSpecError: No compilers for operating system rhel7 satisfy spec gcc@9.3.0
> 
> /gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/package.py:1197, in compiler:
>        1194        """Get the spack.compiler.Compiler object used to build this package"""
>        1195        if not self.spec.concrete:
>        1196            raise ValueError("Can only get a compiler for a concrete package.")
>   >>   1197        return spack.compilers.compiler_for_spec(self.spec.compiler,
>        1198                                                 self.spec.architecture)
> ```
> 
> I would probably solve this by factoring out LLVM and running it after a bootstrap stage to compile GCC.

Interesting.

Thanks for the example.  I'll see if I can reproduce it on my machine.

tldahlgren committed:
Merge remote-tracking branch 'origin/develop' into features/parallel-env-builds

matz-e commented:
> > I'm currently testing this with 12 concurrent processes (each in turn running with `-j 12`) on SLURM, occupying 2 of our compute nodes. Spack, the working directory, and the install directory are all on GPFS 5. I added a delay to my SLURM wrapper so that the build processes start staggered in time (otherwise I got failures that also look like race conditions).
> 
> Good to know. We've seen things like this before with building the provider cache the first time.
> 
> Do you have debug output from where the builds were hanging before you staggered the start time?

They are not really hanging, building seems to go on without adverse effects (that I have noticed so far, still busy patching up build failures). I don't have any debug output right now, but I see messages like this in our Jenkins log, and they disappear with ~1-2 second delay between the individual processes:
```
[2020-11-02T10:22:54.418Z] ### 11:22:54 installing environment
[2020-11-02T10:23:02.656Z] ==> Error: Error writing to config file: '[Errno 2] No such file or directory: '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment/.spack.yaml.tmp' -> '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment/spack.yaml''
[2020-11-02T10:23:02.656Z] ==> Error: Error writing to config file: '[Errno 2] No such file or directory: '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment/.spack.yaml.tmp' -> '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment/spack.yaml''
[2020-11-02T10:23:02.656Z] ==> Installing environment /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment
[2020-11-02T10:23:02.656Z] ==> Error: Error writing to config file: '[Errno 2] No such file or directory: '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment/.spack.yaml.tmp' -> '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment/spack.yaml''
[2020-11-02T10:23:02.656Z] ==> Error: Error writing to config file: '[Errno 2] No such file or directory: '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment/.spack.yaml.tmp' -> '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/merge/deploy/compilers/2020-11-01/data/build_environment/spack.yaml''
```

becker33 commented on lib/spack/spack/cmd/install.py:
```diff
@@ -238,23 +239,29 @@ def default_log_file(spec):
+                with env.write_transaction():
+                    concrete = env.concretize_and_add(abstract, concrete)
+                    env.write(regenerate_views=False)
+            env.install_all(cli_args, **kwargs)
```
If I run
```
spack env create my_project
spack env activate my_project
spack add foo
spack install bar
```
I believe this block of code will install foo as well. Is that the case?

tldahlgren replied:
No, it does not install `foo`.  Should it?

Update:
It will install `bar` since it is provided on the command line.  (It *does* add `bar` to the view.)

If you want to install `foo`, you have to run `spack install` with nothing on the command line.

tgamblin replied:
I verified that this does not happen.

becker33 commented on lib/spack/spack/environment.py:
```diff
@@ -1426,14 +1413,39 @@ def install_all(self, args=None):
+            # Ensure links are set appropriately
+            for spec in specs_to_install:
+                if spec.package.installed:
+                    self._install_log_links(spec)
```
Should this be in a try/except block (still inside the `finally` block) to make sure they all get added even if one fails?

becker33 commented on lib/spack/spack/report.py:
```diff
@@ -44,8 +44,8 @@ def fetch_package_log(pkg):
 class InfoCollector(object):
-    """Decorates PackageInstaller._install_task, which is called by
-    PackageBase.do_install for each spec, to collect information
+    """Decorates PackageInstaller._install_task, which is called via
+    PackageBase.do_install for individual specs, to collect information
```
Does this mean we've lost the ability to do cdash/junit reports on builds of entire environments?

matz-e replied:
Are reports for environments supported currently? In our deployment, [I had to shuffle things around a little in the install command](https://github.com/BlueBrain/spack/commit/411c847189a68266954c90a23e7d9e5a32faceed#diff-5c111ba1f5bfb321d12a190c2a79b33c7b53e3d02f6d0201f1703432b71cb17c) to get JUnit reports to show up, and then take care to not have Spack processes overwrite each others reports. Also seems like the resulting reports will need some de-duplication.

tldahlgren replied:
> Does this mean we've lost the ability to do cdash/junit reports on builds of entire environments?

AFAICT The only thing I ended up changing to this file in this PR is the comment.  

We *used* to collect data based on the package `do_install` method because the build would call each dependency package's `do_install` (so their data would be collected).  

The distributed build work involved a refactor to use a build queue.  This basically resulted in individual packages getting installed by the `_install_task`, not `do_install` method, in order to continue to collect the data on dependency packages.  Hence the switch described in the comment.

I haven't tested report collection on environments so I'll try that.

tldahlgren replied:
> Are reports for environments supported currently? In our deployment, [I had to shuffle things around a little in the install command](https://github.com/BlueBrain/spack/commit/411c847189a68266954c90a23e7d9e5a32faceed#diff-5c111ba1f5bfb321d12a190c2a79b33c7b53e3d02f6d0201f1703432b71cb17c) to get JUnit reports to show up, and then take care to not have Spack processes overwrite each others reports. Also seems like the resulting reports will need some de-duplication.

Interesting.  Looks like we might need a bit of your shuffle added to the main line.

tldahlgren replied:
Confirmed that the report does not appear to be generated.

tldahlgren replied:
> > Are reports for environments supported currently? In our deployment, [I had to shuffle things around a little in the install command](https://github.com/BlueBrain/spack/commit/411c847189a68266954c90a23e7d9e5a32faceed#diff-5c111ba1f5bfb321d12a190c2a79b33c7b53e3d02f6d0201f1703432b71cb17c) to get JUnit reports to show up, and then take care to not have Spack processes overwrite each others reports. Also seems like the resulting reports will need some de-duplication.
> 
> Interesting. Looks like we might need a bit of your shuffle added to the main line.

@matz-e I added your "shuffling" in my latest commit (0d7ce1f) and confirmed at least the `junit` output is now generated for an environment.  Thank you!

matz-e replied:
Thanks, glad it's of some use!

becker33 commented on lib/spack/spack/installer.py:
```diff
@@ -544,61 +587,17 @@ def package_id(pkg):
-                suppresses it)
-        """
+
+    return "{0}-{1}-{2}".format(pkg.name, pkg.version, pkg.spec.dag_hash())
```
Yay!

becker33 commented on lib/spack/spack/installer.py:
```diff
@@ -1802,6 +2035,15 @@ def __init__(self, message, long_msg=None):
+
+    def __init__(self, pkg_name, phase):
+        super(BadInstallPhase, self).__init__(
+            '\'{0}\' is not an allowed phase for package {1}'
```
```suggestion
            '\'{0}\' is not an valid phase for package {1}'
```

becker33 reviewed (dismissed):
I'll probably have to go over this again since some of the logic is pretty hard to reason about, but it mostly looks like a reasonable architecture and I've noted everything I noticed this time through.

tldahlgren committed:
Update lib/spack/spack/installer.py

Adjust exception wording

Co-authored-by: Greg Becker <becker33@llnl.gov>

tldahlgren committed:
Ensure bad phase message consistent with suggested change from previous commit

tldahlgren committed:
Put _install_log_links in try-except block

tldahlgren committed:
Resolve test_cache_install_full_hash_match failure

tldahlgren committed:
Restore install reporting (thanks matz-e!)

matz-e commented on lib/spack/spack/installer.py:
```diff
@@ -1489,11 +1520,9 @@ def install(self, **kwargs):
-                    _print_installed_pkg(pkg.prefix)
+            if not task.explicit:
+                if _handle_external_and_upstream(pkg, False):
+                    self._flag_installed(pkg, get_dependent_ids(spec))
```
I had some issues when, e.g., an external `cmake` showed up several times in separately concretized specs, and it was not removed from all dependents' uninstalled dependencies. This fixed said scenario with my small reproducer (using our internal packages, unfortunately):
```suggestion
                    self._flag_installed(pkg, task.dependents)
```
At least here, I think it's better to flag for the superset of dependents (which is also readily available), rather than just the ones from one of the specs in the tree.

tldahlgren replied:
> I had some issues when, e.g., an external `cmake` showed up several times in separately concretized specs, and it was not removed from all dependents' uninstalled dependencies. This fixed said scenario with my small reproducer (using our internal packages, unfortunately):
> 
> At least here, I think it's better to flag for the superset of dependents (which is also readily available), rather than just the ones from one of the specs in the tree.

Good point, thanks!

Reviewing other calls to `_flag_installed` makes me realize there's a small, related refactor that should make the processing being done here a bit clearer.

tldahlgren committed:
Use task dependents to flag non-locally installed dependents PLUS refactor _flag_installed for clarity

tldahlgren committed:
Merge remote-tracking branch 'origin/develop' into features/parallel-env-builds

tldahlgren committed:
Attempt to fix MacOS test_env_install_two_specs_same_dep test failure

tldahlgren committed:
Added test_push_task_skip_processed

tgamblin committed:
fix Python 3.8 test errors

tgamblin commented on lib/spack/spack/environment.py:
```diff
@@ -1402,19 +1386,22 @@ def _install(self, spec, **install_args):
 
+        Args:
+            args (Namespace): argparse namespace with command arguments
+            install_args (dict): keyword install arguments
```
It's a bit confusing to have both args and `install_args` -- this should probably be refactored at some point.

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -1782,21 +1873,163 @@ def flag_installed(self, installed):
+        """
+        # Ensure dealing with a package that has a concrete spec
+        if not isinstance(pkg, spack.package.PackageBase):
+            raise ValueError("{0} must be a package".format(str(pkg)))
```
Should be a `TypeError`

healther commented:
should this work out of the box? at least for me on WSL2 within an environment it does not install multiple packages in parallel, but only does the normal make-based parallelisation. I'm unsure if this should work just as is, or I have to enable it manually in some config and I didn't manage to find out via the documentation so I thought I just ask here

matz-e commented:
How do you launch it? I can easily trigger build parallelism by launching the same command `spack -D ./my_env_dir install` in two or more terminals (depending on how many steps of the dependency DAG can be parallelized).

tldahlgren commented:
> should this work out of the box? at least for me on WSL2 within an environment it does not install multiple packages in parallel, but only does the normal make-based parallelisation. I'm unsure if this should work just as is, or I have to enable it manually in some config and I didn't manage to find out via the documentation so I thought I just ask here

If by "out of the box" you mean it will launch multiple processes to run the build in parallel, then the answer is "no".  The jobs option (`-j`) continues to be passed to/used by packages whose builds support that option (e.g., make).

As `matz-e` points out, support for parallel builds -- including of environments -- involves coordination between multiple processes for the same Spack instance.

Which means you can launch multiple `spack install` commands (e.g., in the background on a single node) or launch them in batch jobs on multiple processes/nodes.  The goal is for separate processes to coordinate the installation of the packages represented by the dependency DAG (or DAGs in the case of multiple specs such as defined by an environment).

The effective number of coordinating processes for a given environment or spec is the maximum number of packages in the dependency DAG(s) that have no uninstalled dependencies.

healther commented:
Ah so I expected:
```
spack env activate general
spack concretize -f
spack install
```
to spawn as many (at least number of top-level specs many) processes in order to parallelise the installation. What I saw is that the typical utilisation is 1 core and during normal `make`ish steps it uses the whole 8 cores.

If I understand you correctly then I should do
```
spack env activate general
spack concretize -f
spack install &
spack install & 
[...]
```
correct @matz-e ?

I was hoping for a more automated process (and actually understood the discussion here in that light, but I only skimmed over it)

matz-e commented:
@healther correct, as @tldahlgren points out.

I have my own wrapper for our deployment, which also ensures that JUnit reports end up in unique files. It's certainly a bit more work than having everything built in, but I appreciate the flexibility to launch building across several nodes of our cluster.

healther commented:
I see, yeah for cluster management I think I see the appeal of doing it explicitly, but for me just rebuilding my own stuff locally "batteries included" would be more convenient ;) thanks for the feedback

haampie commented on lib/spack/spack/cmd/install.py:
```diff
@@ -294,8 +305,17 @@ def install(parser, args, **kwargs):
-            env.install_all(args)
+            specs = env.all_specs()
+            if not args.log_file and not reporter.filename:
+                reporter.filename = default_log_file(specs[0])
```
@tldahlgren environments don't always have a nonzero number of specs, and then `spack install` fails. Could you potentially fix this? I'm not sure what `reporter` is.

Currently on an env with `specs: []`:

```
$ spack install -v
==> Error: list index out of range
```

which is thanks to `specs[0]` in this line ^

tldahlgren replied:
There's a quick change (or fix?) at #28031
